lib/Remotery.c

```diff
@@ -10091,7 +10091,8 @@ typedef struct VulkanBindImpl
     VulkanSample** sampleRingBuffer;
 
     // Read/write positions of the ring buffer allocator, synchronising access to all the ring buffers at once
-    // TODO(don): Separate by cache line?
+    // NOTE(valakor): These are 64-bit instead of 32-bit so that we can reasonably assume they never wrap.
```

Comment:
Why is this and is there any code that will break if these positions wrap?

(is it the reference to ABA?)

Reply:
If these counters wrap then a whole bunch of existing code that does comparisons like `current_write_gpu > current_read_cpu` breaks. It could be rewritten with some amount of effort (maybe), but using 64bit counters avoids the problem in the first place (for any reasonable program).

For example, a program running at 1000fps that writes 200 timestamps per frame will wrap a 32-bit counter in ~6 hours, whereas a 64-bit counter won't wrap for almost 3 million years.

The D3D12 code still uses 32-bit counters but could be modified without too much effort if we wanted to.

Reply:
Also, Vulkan timeline semaphores are 64-bit counters that must strictly increase so they're basically not allowed to wrap.

Reply:
The culprit is `VulkanMarkFrame` (and `D3D12MarkFrame`) where there are comparisons between `current_write_gpu`, `current_read_cpu`, and `msg_sample_tree->userData`.

That aside, Vulkan still requires strictly increasing values for semaphores. I guess this could be avoided by creating a new semaphore when things are about to wrap but that feels more complicated than just using 64-bit counters.

Reply:
OK, that's a bug! I will either fix that or move the D3D12 ones to 64-bit indices.